src/qhttp/LogHelpers.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * LSST Data Management System
+ * Copyright 2022 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_QSERV_QHTTP_LOGHELPERS_H
+#define LSST_QSERV_QHTTP_LOGHELPERS_H
+
+// System headers
+#include <iostream>
+#include <memory>
+
+// Local headers
+#include "qhttp/AjaxEndpoint.h"
+#include "qhttp/Server.h"
+
+// Third-party headers
+#include "boost/asio.hpp"
+
+namespace lsst {
+namespace qserv {
+namespace qhttp {
+
+//
+//----- Helper ouput stream manipulators for logging.  These let you say things like:
+//      logger(self) << logger(sock) << "log message..." in logger macro calls
+//
+
+struct ServerLogger
+{
+    Server *server;
+    ServerLogger(Server *server) : server(server) {};
+    friend std::ostream& operator<<(std::ostream &str, ServerLogger logger) {
```

Comment:
Why `friend`? This is not needed because members of the `struct` are already public.

Reply:
I had been following standard output manipulator pattern here.  In this case, though, since struct members are public, the operator specialization could be implemented without `friend` if moved out of the struct def'n and declared `inline`.  That is only slightly more verbose, but I can make that change here if you think it is stylistically preferable?

Reply:
The `friend` is only needed for the operator to access protected or private members (data or methods) of the class. Hence the pattern you're referring to. In case if you have `struct` you simply move the operator definition out of the class as:
```
struct ServerLogger
{
    Server *server;
    ServerLogger(Server *server) : server(server) {};
};
std::ostream& operator<<(std::ostream &str, ServerLogger logger) { ... }
```
My second comment is that since `ServerLogger` is not a primitive data type then you want to pass it into the operator by `const&` reference:
```
std::ostream& operator<<(std::ostream &str, ServerLogger const& logger) ...
```

Reply:
Yes, exactly; see comment above.  If you would prefer to see operator defined outside the struct this way, I am happy to change this -- it's only an extra line or two per manipulator.

Reply:
acknowledged on `const&` here -- will do

Reply:
We've been told that `friend` is evil in C++, pretty much like `goto` 😄 That's why it caught my attention. Besides, there are always ways to avoid the "forced friendship", for example by implementing the public method `toString()` like in:
```
class MyClass {
public:
    std::string toString() const;
private:
    int _foo;
    std::string _bar;
};
std::ostream& operator<<(std::ostream &str, MyClass const& obj) {
    str << obj.toString();
    return str;
} 
```
A GOD think about this approach is that the string serialization logic is now a part of the class, not a separate piece of code that would have to be kept in SYNC sith the class should the class implementation evolve over time (new data members, removed data members, etc.).

Reply:
By the way, why do you need `ServerLogger`? You might as well overload the operator `<<` for the target class `Server`. In your implementation, the class `ServerLogger` is used solely to store a pointer to `Server` and to have the overloaded operator ... that in the end would go straight to the stored pointer. What are the benefits for this extra layer of indirection?

Reply:
The benefit is being able to say `log_ << logger(server) << "something happened with the server"` and have that result in a log-specific presentation of the server.   Also, `logger()` can be made to work on both regular pointers and smart pointers without having to worry which is which.  The output we want for the log wrt. one of these objects is not necessarily what we generally would want for output streaming (for example, has "srv=" for context, hex addr pointers, etc., logs identity and not value, etc. etc.)

Reply:
In any case, I'll go with "the kids say `friend` is evil" -- not wanting to be all boomer about it, I will rewrite the operator overloads outside the struct def'n :-)